Sake/Mixer.cs

```diff
@@ -256,62 +256,21 @@ public IEnumerable<ulong> Decompose(IEnumerable<ulong> myInputsParam, IEnumerabl
                 (naiveSet, loss + (ulong)naiveSet.Count * OutputFee)); // The cost is the remaining + output cost.
 
             // Create many decompositions for optimization.
-            var before = DateTimeOffset.UtcNow;
-            do
+            Decomposer.StdDenoms = denoms.Where(x => x <= myInputSum).Select(x => (long)x).ToArray();
+            foreach (var (sum, count, decomp) in Decomposer.Decompose((long)myInputSum, (long)MinAllowedOutputAmountPlusFee, Math.Max(3, naiveSet.Count)))
             {
-                var currSet = new List<ulong>();
-                remaining = myInputs.Sum();
-                do
-                {
-                    var denomPlusFees = denoms.Where(x => x <= remaining && x >= (remaining / 3)).ToList();
-                    if (!denomPlusFees.Any())
-                    {
-                        break;
-                    }
-
-                    var denomPlusFee = denomPlusFees.RandomElement();
-                    if (remaining < MinAllowedOutputAmountPlusFee)
-                    {
-                        break;
-                    }
-
-                    if (denomPlusFee <= remaining)
-                    {
-                        currSet.Add(denomPlusFee);
-                        remaining -= denomPlusFee;
-                    }
-                }
-                while (currSet.Count <= naiveSet.Count || currSet.Count <= 3);
+                var set = Decomposer.ToRealValuesArray(
+                    decomp,
+                    count,
+                    Decomposer.StdDenoms).Cast<ulong>().ToArray();
 
-                // If currSet.Count <= 3 then we still generate sets to add ambiguity. 
-                if (currSet.Count <= naiveSet.Count || currSet.Count <= 3)
+                hash = new();
+                foreach (var item in naiveSet.OrderBy(x => x))
```

Comment:
@nopara73 You are computing always the hashcode corresponding to the naive set and that's the problem.

Reply:
Thanks!